vms/platformvm/docs/validators_versioning.md

```diff
@@ -0,0 +1,113 @@
+
+In contrast, `BLS Public Key` diffs are *backward-looking*: a diff recorded at height `H` stores the change that transforms validator `BLS Public Key` at height `H` into validator `BLS Public key` at height `H-1`.
+
+Finally, if no changes are made to the validator set no diff entry is recorded. This implies that a validator `Weight` or `BLS Public Key` diff may not be stored for every height `H`.
+
+## Validator diffs layout
+
+Validator diffs layout is optimized to support iteration. Validator sets are rebuilt by accumulating `Weight` and `BLS Public Key` diffs from the top-most height down to the requested height. So validator diffs are stored so that it's fast to iterate them in this order.
+
+`Weight` diffs are stored as a contiguous block of key-value pairs as follows:
+
+| Key                                | Value                                |
+|------------------------------------|--------------------------------------|
+| SubnetID + Reverse_Height + NodeID | serialized state.ValidatorWeightDiff |
+
+Note that:
+
+1. `Weight` diffs related to a Subnet are stored contiguously.
+2. Diff height is serialized as `Reverse_Height`. It is stored with big endian format and has its bits flipped too. Big endianess ensures that heights are stored in order, bit flipping ensures that the top-most height is always the first.
+3. `NodeID` is part of the key and `state.ValidatorWeightDiff` is part of the value.
+
+`BLS Public` diffs are stored as follows:
+
+| Key                                | Value                         |
+|------------------------------------|-------------------------------|
+| SubnetID + Reverse_Height + NodeID | validator.BlsKey bytes or nil |
+
+Note that:
+
+1. `BLS Public Key` diffs have the same keys as `Weight` diffs. This implies that the same ordering is guaranteed.
+2. Value is either validator `BLS Public Key` bytes or an empty byte slice, as illustrated in previous section.
+
+## Validators diff usage in rebuilding validators state
+
+Now let's see how diffs are used to rebuild the validator set at a given height. The procedure varies slightly between Primary Network and Subnet validator, so we'll describe them separately.
+We assume that the reader knows that, as of the Cortina fork, every Subnet validator must also be a Primary Network validator.
+
+### Primary network validator set rebuild
+
+If the P-Chain's current height is `T` and we want to retrieve the Primary Network validators at height `H < T`. We proceed as follows:
+
+1. We retrieve the Primary Network validator set at current height `T`. This is the base state on top of which diffs will be applied.
+2. We apply weight diffs first. Specifically:
+   -  `Weight` diff iteration starts from the top-most height smaller or equal to `T`. Remember that entries heights do not need to be contiguous, so the iteration starts from the highest height smaller or equal to `T`, in case `T` has not diff entry.
```

Comment:
```suggestion
   -  `Weight` diff iteration starts from the top-most height smaller or equal to `T`. Remember that entry heights do not need to be contiguous, so the iteration starts from the highest height smaller or equal to `T`, in case `T` does not have a diff entry.
```